wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,393 @@
+
+-- | Wallet Submission Layer
+--
+-- This module implements section 9 of the Wallet spec,
+-- namely 'Transaction Submission'.
+--
+data WalletSubmission m = WalletSubmission {
+      _wsResubmissionFunction :: ResubmissionFunction m
+    , _wsState                :: WalletSubmissionState
+    }
+
+instance Buildable (WalletSubmission m) where
+    build ws = bprint ("WalletSubmission <rho> " % F.build) (_wsState ws)
+
+data WalletSubmissionState = WalletSubmissionState {
+      _wssPendingSet  ::  Pending
+    , _wssScheduler   ::  Scheduler
+    , _wssCurrentSlot :: !Slot
+    }
+
+instance Buildable WalletSubmissionState where
+    build wss = bprint ("{ pendingSet = " % F.build %
+                        ", scheduler  = " % F.build %
+                        ", slot       = " % F.build %
+                        " } "
+                       ) (_wssPendingSet wss) (_wssScheduler wss) (_wssCurrentSlot wss)
+
+-- | A 'Scheduler'. Despite modelled as in 'IntMap' it has to be intended
+-- as a mapping between 'Slot' and the list of transactions due that slot.
+newtype Scheduler = Scheduler {
+    _ssScheduled :: IntMap [Scheduled]
+    }
+
+newtype Scheduled = Scheduled (Core.TxId, Core.TxAux, Schedule) deriving Eq
+
+instance Buildable Scheduled where
+    build (Scheduled (txId, _, s)) = bprint pairF (txId, s)
+
+instance Buildable [Scheduled] where
+    build s = bprint (listJsonIndent 4) s
+
+-- | Our \"opaque\" concept of 'Slot', which might or might not line up with
+-- the 'Core.FlatSlotId' of the blockchain.
+-- Modelled as an 'Int' to tap into the performance of things like 'IntMap's,
+-- and enough to keep a ticker running for hundreds of years. Remember this is
+-- not the lifetime of the blockchain: it has more to do with the lifetime of
+-- the wallet, as it will reset to 0 each time we restart it (the entire
+-- 'WalletSubmission' is ephimeral and not persisted on disk).
+type Slot = Int
+
+-- | How many times we have tried to submit the given transaction.
+-- When this value reaches the 'maxRetries' value, the transcation will be
+-- removed from the local pending set.
+-- Note that when the @Core@ layer will introduce the concept of \"Time to
+-- Live\" for transactions, we will be able to remove the 'maxRetries' value
+-- and simply use the @TTL@ to judge whether or not we should retry.
+type SubmissionCount = Int
+
+-- | A default and very generous time to live.
+maxRetries :: SubmissionCount
+maxRetries = 255
+
+-- | The 'Schedule' for a 'Core.TxAux'.
+newtype Schedule = Schedule SubmissionCount deriving Eq
+
+instance Buildable Schedule where
+    build (Schedule s) = bprint F.build s
+
+
+-- | The 'Evicted' set represents the transactions which needs to be
+-- pruned from the local (and wallet) 'Pending' set.
+type Evicted = Pending
+
+-- | A 'ResubmissionFunction' (@rho@ in the spec), parametrised by an
+-- arbitrary @m@.
+type ResubmissionFunction m =  Slot
+                            -- ^ The current slot
```

Comment:
Ah, I see (by reading the later code). Since this isn't a list anymore, this is the only way that we can say "insert 10 slots from now". Might be useful to mention this.